### PR TITLE
Check Type has parameters field with an element

### LINKED ISSUE #491: Some Types don't have field parameters error when checking if annotation is of form "String::Type{String}"
```jl
using StrideArraysCore, StrideArrays

julia> @inline function smallarray4(::Type{T}, ::StaticInt{SpillSize}, size::Integer...) where {T, SpillSize}
           @assert isbitstype(T)
           N = prod(size)
           if N <= SpillSize
               data = Ref{NTuple{SpillSize, T}}()
               p = Ptr{T}(pointer_from_objref(data))
           else
               data = Vector{T}(undef, N)
               p = Ptr{T}(pointer(data))
           end
           PtrArray(p, size, Vector{T}(undef, 32))
       end
smallarray4 (generic function with 1 method)

julia> @descend smallarray4(Float32, static(16), 16)
smallarray4(::Type{T}, ::StaticInt{SpillSize}, size::Integer...) where {T, SpillSize} @ Main REPL[210]:1
┌ Warning: Some line information is missing, type-assignment may be incomplete
└ @ Cthulhu ~/.julia/packages/Cthulhu/DxawF/src/codeview.jl:144
ERROR: type TypeofBottom has no field parameters
Stacktrace:
  [1] getproperty
    @ Base ./Base.jl:32 [inlined]
  [2] type_annotation_mode(node::JuliaSyntax.TreeNode{…}, T::Any; type_annotations::Bool, hide_type_stable::Bool)
    @ TypedSyntax ~/.julia/packages/TypedSyntax/ByLzZ/src/show.jl:104
  [3] type_annotation_mode
    @ TypedSyntax ~/.julia/packages/TypedSyntax/ByLzZ/src/show.jl:97 [inlined]
  [4] printstyled(io::IOContext{…}, rootnode::JuliaSyntax.TreeNode{…}; type_annotations::Bool, iswarn::Bool, hide_type_stable::Bool, with_linenumber::Bool, idxend::Int64)
    @ TypedSyntax ~/.julia/packages/TypedSyntax/ByLzZ/src/show.jl:45
  [5] cthulhu_typed(io::IOContext{…}, debuginfo::Symbol, src::Core.CodeInfo, rt::Any, effects::Core.Compiler.Effects, mi::Core.MethodInstance; iswarn::Bool, hide_type_stable::Bool, optimize::Bool, pc2remarks::Nothing, pc2effects::Nothing, inline_cost::Bool, type_annotations::Bool, annotate_source::Bool, inlay_types_vscode::Bool, diagnostics_vscode::Bool, interp::Cthulhu.CthulhuInterpreter)
    @ Cthulhu ~/.julia/packages/Cthulhu/DxawF/src/codeview.jl:168
  [6] cthulhu_typed(::IOContext{…}, ::Cthulhu.DInfo.DebugInfo, ::Core.CodeInfo, ::Vararg{…}; kwargs::@Kwargs{…})
    @ Cthulhu ~/.julia/packages/Cthulhu/DxawF/src/codeview.jl:121
  [7] (::Cthulhu.var"#92#99"{…})(lambda_io::IOContext{…})
    @ Cthulhu ~/.julia/packages/Cthulhu/DxawF/src/Cthulhu.jl:481
  [8] stringify(f::Any, context::IOContext{Base.TTY})
    @ Cthulhu ~/.julia/packages/Cthulhu/DxawF/src/ui.jl:90
  [9] 
    @ Cthulhu ~/.julia/packages/Cthulhu/DxawF/src/Cthulhu.jl:480
 [10] _descend(term::REPL.Terminals.TTYTerminal, interp::Cthulhu.CthulhuInterpreter, mi::Core.MethodInstance; kwargs::@Kwargs{…})
    @ Cthulhu ~/.julia/packages/Cthulhu/DxawF/src/Cthulhu.jl:739
 [11] _descend(term::REPL.Terminals.TTYTerminal, args::Any; interp::Core.Compiler.NativeInterpreter, kwargs::@Kwargs{…})
    @ Cthulhu ~/.julia/packages/Cthulhu/DxawF/src/Cthulhu.jl:755
 [12] __descend_with_error_handling(args::Any; terminal::Any, kwargs...)
    @ Cthulhu ~/.julia/packages/Cthulhu/DxawF/src/Cthulhu.jl:211
 [13] _descend_with_error_handling(f::Any, argtypes::Any; kwargs::@Kwargs{iswarn::Bool})
    @ Cthulhu ~/.julia/packages/Cthulhu/DxawF/src/Cthulhu.jl:200
 [14] descend_code_typed(::Any, ::Vararg{Any}; kwargs::@Kwargs{})
    @ Cthulhu ~/.julia/packages/Cthulhu/DxawF/src/Cthulhu.jl:158
 [15] top-level scope
    @ REPL[216]:1
Some type information was truncated. Use `show(err)` to see complete types.
```
Probably related to the function throwing
```jl
julia> smallarray4(Float32, static(16), 16)
ERROR: MethodError: no method matching (PtrArray)(::Ptr{Float32}, ::Tuple{Int64}, ::Vector{Float32})

Closest candidates are:
  (PtrArray)(::Ptr{T}, ::S, ::Val{R}) where {T, N, S<:Tuple{Vararg{Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8, StaticInt}, N}}, R}
   @ StrideArraysCore ~/.julia/packages/StrideArraysCore/COJRJ/src/ptr_array.jl:307
  (PtrArray)(::Ptr{T}, ::Tuple{Vararg{Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8, StaticInt}, N}}, ::Tuple{Vararg{Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8, StaticInt}, N}}, ::Val{D}) where {T, N, D}
   @ StrideArraysCore ~/.julia/packages/StrideArraysCore/COJRJ/src/ptr_array.jl:556
  (PtrArray)(::Ptr{T}, ::S, ::X, ::Val{R}) where {T, N, S<:Tuple{Vararg{Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8, StaticInt}, N}}, X<:Tuple{Vararg{Any, N}}, R}
   @ StrideArraysCore ~/.julia/packages/StrideArraysCore/COJRJ/src/ptr_array.jl:325
```

### PR DESCRIPTION
**Pull request by Zentrik**
Fixes #491, #492, #494

### CODE REVIEW

**Comment by Zentrik** [comment]
Unfortunately it seems that `extract_inner_type(::Type{Type{T}})` specializes on each new type it sees, I don't know how problematic that is. Also not sure why we were taking the first element of `T.parameters` before but this implementation seems to work fine.

File: TypedSyntax/src/show.jl [hunk 1]
```diff
@@ -94,17 +94,17 @@ function is_show_annotation(@nospecialize(T); type_annotations::Bool, hide_type_
...
 end
 
+extract_inner_type(x) = nothing
+extract_inner_type(::Type{Type{T}}) where T = T
 function type_annotation_mode(node, @nospecialize(T); type_annotations::Bool, hide_type_stable::Bool)
     kind(node) == K"return" && return false, "", "", ""
     type_annotate = is_show_annotation(T; type_annotations, hide_type_stable)
     pre = pre2 = post = ""
     if type_annotate
-        if T isa Type && T <: Type
```
Comment (timholy) [inline TypedSyntax/src/show.jl:102]: Hmm, I guess this should have been `T <: Type{S} where S` or something?

**Review by timholy — COMMENTED** [review]
I doubt it will be measurable (and so we can merge this), but presumably `extract_inner_type` will be specialized for every type. The old code tried to avoid that but it may not be readily fixable. So I'm fine with merging this, thanks for tackling it!

**Comment by timholy** [comment]
> Also not sure why we were taking the first element of `T.parameters` before but this implementation seems to work fine.

```julia
julia> TT = Type{Float64}
Type{Float64}

julia> TT.parameters[1]
Float64
```

Basically we were extracting the `T` in the same way as your `extract_inner_type`, without forcing dispatch specialization. I think the problem with the old implementation is that `T <: Type` was too weak of a condition. Perhaps `T <: Type{S} where S` might fix it?

File: TypedSyntax/src/show.jl (hunk 1 above)
Comment (Zentrik) [reply TypedSyntax/src/show.jl:102]: `Union{} <: Type{S} where S` is true so I don't think that would have fixed it either.

**Comment by Zentrik** [comment]
I thought `extract_inner_type(@nospecialize(x::Type{<:Type})) = x.parameters[1]` might work, but unfortunately it fails with `Core.TypeofBottom`. Given I don't know when a type actually has a non empty `parameters` `svec` might be easiest to just use this implementation.

**Comment by simeonschaub** [comment]
I believe it should be enough to just change the check to `T isa DataType && T <: Type`

**Comment by Zentrik** [comment]
`Core.TypeofBottom` satisfies that but has an empty parameters svec, maybe checking if it's non empty as well will work.

**Comment by Zentrik** [comment]
Given we don't think the current implementation will be particularly slow I'll merge it for now, and we can revisit this if it is slow.

File: TypedSyntax/src/show.jl (hunk 1 above)
Comment (vtjnash) [reply TypedSyntax/src/show.jl:102]: It should be `T isa DataType` or `Base.isType(T)`

**Pull request merged**